### ISSUE TITLE
fsck and mount systemd units fail on multidevice bcachefs filesystem

### ISSUE THREAD

**Issue by ZoomRmc** [NixOS/nixpkgs#72970]
**Describe the bug**
fsck systemd unit fails on multidevice bcachefs filesystems with
`Timed out waiting for device /dev/sda1:/dev/sdb1:/dev/sdc1`, thus failing the mount unit being a dependency.
The issue, as I understand, is that `bcachefs mount` expects the used partitions to be separated by colon, while fsck needs them separated by whitespace.

**To Reproduce**
1. Create multidevice bcachefs filesystem 
`bcachefs format /dev/sd[ab]1`
2. Add it to your hardware-configuration.nix and rebuild.
`"/mnt" = { device = "/dev/sda1:/dev/sdb1";  fsType = "bcachefs";  };`

**Expected behavior**
Units not failing.

**Metadata**
 - version: `nix-env (Nix) 2.3pre6895_84de821`
 - channels(root): `"nixos-20.03pre200231.7827d3f4497"`

**Comment by ZoomRmc** [comment]
Still important to me.

**Comment by wucke13** [comment]
And for me as well.

**Comment by bqv** [comment]
Ditto.

**Comment by ZoomRmc** [comment]
Good bot.

**Comment by wucke13** [comment]
Related issue on systemd: https://github.com/systemd/systemd/issues/8234 btw.

**Comment by Madouura** [comment]
Is this still relevant?
I don't have any problems anymore on my end, and ``nixos-generate-config`` correctly generates what should be mounted, even multi-device, and it seems like it mounts properly.
AFAIK bcachefs does it's own fsck on mount, but I could be wrong.

**Comment by wucke13** [comment]
@Madouura Just to be clear, you are booting of a multi-device bcachefs root device?

**Comment by Madouura** [comment]
/boot itself is an ef00 partition with a fat filesystem, but root is multi-device bcachefs, encrypted too.

**Comment by Madouura** [comment]
```nix
  fileSystems."/" =
    { device = "/dev/sda1:/dev/nvme0n1p2:/dev/nvme1n1p1";
      fsType = "bcachefs";
    };

  fileSystems."/boot" =
    { device = "/dev/disk/by-uuid/5BCE-E0D5";
      fsType = "vfat";
    };
```

**Comment by wucke13** [comment]
Very nice, this seems to indicate my side of the problem was fixed :)

**Comment by Madouura** [comment]
Is there any reason to keep the issue open, or am I missing something then?

**Comment by Madouura** [comment]
Should be fine to close, as it seems to be fixed. If it needs to be re-opened then it can be.

**Comment by ZoomRmc** [comment]
The problem still persists, although, I'm not sure the reasons are still the same - currently, NixOS can't generate a fsck service name.
I need to be using persistent block device names, as I have multiple controllers and short labels change randomly on boot.

```
[   20.178828] systemd-fstab-generator[492]: Failed to create fsck service name: File name too long
[   21.176951] systemd[488]: /nix/store/kxqqbyxf4w0bg4n2ip1qq3kr5bw4hdq0-systemd-249.7/lib/systemd/system-generators/systemd-fstab-generator failed with exit status 1.
```

The config is:
```nix
    "/data" = {
      device = "/dev/disk/by-id/wwn-0x50014ee2671c3970:/dev/disk/by-id/wwn-0x50014ee657c4ef16-part1:/dev/disk/by-id/wwn-0x50014ee657dcf49d-part3:/dev/disk/by-id/wwn-0x5000cca369ce34a9:/dev/disk/by-id/wwn-0x5000039ffef41ed1";
      fsType = "bcachefs";
      options = [ "verbose" "nofail" "noatime" "x-systemd.device-timeout=25s" ];
    };
```

Do I need to file a separate issue?

**Comment by Madouura** [comment]
Nah, I must have misunderstood the issue.
Let's go ahead and reopen.

**Comment by Slabity** [comment]
Oddly I am actually still running into the original issue even on unstable. If I add the following:

```nix
    fileSystems."/data" = {
      device = "/dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1";
      fsType = "bcachefs";
    };
```

Then systemd times out with `Timed out waiting for device /dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1` and then fails to boot.

**Comment by firestack** [comment]
@Slabity Are you having any other issues than just timeout on load? from your comment and log message it doesn't seem like your unit is having trouble with the device string but rather timing out due to having to walk the full journal, Have you tried adding `x-systemd.device-timeout=`? I've got mine set to 2m (for worse case after improper shutdown) for a 12TB array.

**Comment by Madouura** [comment]
>/dev/md127p1

I'm not sure bcachefs supports another RAID(?) device in this string to
begin with.
What is this device?

**Comment by Slabity** [comment]
> /dev/md127p1
> I'm not sure bcachefs supports another RAID(?) device in this string to begin with. What is this device?

It's a RAID5 MD device with 3 8TB HDDs that I use as the background target. I used MD because bcachefs doesn't seem to have RAID5 support yet. I have no issues when setting it up and I can mount it manually with no issue. It's only at boot time when systemd times out.

> @Slabity Are you having any other issues than just timeout on load? from your comment and log message it doesn't seem like your unit is having trouble with the device string but rather timing out due to having to walk the full journal, Have you tried adding x-systemd.device-timeout=? I've got mine set to 2min (for worse case after improper shutdown) for a 12TB array.

Is walking the journal part of fsck? If so, I don't think it's getting to that yet. When the timeout occurs it ends up dropping me into an emergency shell. From there I check `journalctl -xb` and see the following:

```
May 14 23:35:40 nixos systemd[1]: dev-nvme1n1p1:-dev-nvme2n1p1:-dev-md127p1.device: Job dev-nvme1n1p1:-dev-nvme2n1p1:-dev-md127p1.device timed out.
May 14 23:35:40 nixos systemd[1]: Timed out waiting for device /dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1.
Subject: Unit dev-nvme1n1p1:-dev-nvme2n1p1:-dev-md127p1.device has failed
Defined-By: systemd
Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel

Unit dev-nvme1n1p1:-dev-nvme2n1p1:-dev-md127p1.device has failed

The result is RESULT
May 14 23:35:40 nixos systemd[1]: Dependency failed for File System Check on /dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1.
Subject: Unit systemd-fsck@dev-nvme1n1p1:-dev-nvme2n1p1:-dev-md127p1.service has failed
Defined-By: systemd
Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel

Unit systemd-fsck@dev-nvme1n1p1:-dev-nvme2n1p1:-dev-md127p1.service has failed

The result is RESULT
```

Which I believe means it's not even getting to the `fsck` checks because it's not finding the device. 

I can also confirm earlier in the logs that md127 is seen and exists, so I'm confident I have my MD modules setup correctly and all 3 devices exist at the time. Once I'm in the emergency shell I can run the following and it mounts instantly:

`mount -t bcachefs /dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1 /data`

I will try playing with different options, but I'm a little confused how this is working for anyone at all. According to the systemd issue linked, it doesn't support multiple devices yet and some workaround with a custom mount service is required. Unfortunately I have no idea what a service file for that would even look like.

**Comment by Madouura** [comment]
I'm guessing it's failing in this block, likely due to it not being able to check the RAID device. https://github.com/NixOS/nixpkgs/blob/34e4df55664c24df350f59adba8c7a042dece61e/nixos/modules/system/boot/stage-1-init.sh#L85
I don't know why it can't just time out and then mount anyway though, that's confusing.

**Comment by Slabity** [comment]
> I'm guessing it's failing in this block, likely due to it not being able to check the RAID device.

Hmm... Well this is happening in stage 2. In fact if I add the "nofail" option then my system boots up fine and I can log in before it even finishes timing out.

**Comment by firestack** [comment]
I'm pretty sure the mount time scales with the amount of data bcachefs needs to read before it can mount it completely. As there's some sort of in memory structure that needs to be built currently (this is planned to be addressed)

(https://www.patreon.com/posts/9293694)
> On larger filesystems, bcachefs's mount times still are too slow - this is really only a stopgap measure until I implement persistent allocation information and a few other things. Fsck performance appears to be quite good compared to other filesystems, though. (Could use benchmarks if anyone wants to run them).

If you want to know what `bcachefs` is doing and if it's stuck somewhere, you should set `options = ["verbose"];`
https://github.com/koverstreet/bcachefs/issues/318#issuecomment-932264367

**Comment by Slabity** [comment]
@firestack  - Thanks, but I'm pretty sure it is not the raid device or bcachefs taking too long. This is specifically systemd being unable to find the device.

I can run `sudo systemctl start data.mount` after logging in and it still will timeout with the same `Timed out waiting for device /dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1` error. On the other hand I can run `sudo mount -t bcachefs /dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1 /data` and it will mount in 2 seconds.

It looks like we do something special in stage-1 by manually splitting the devices to get them to work. I'm not sure we do that in stage-2 or when creating the mount service or wherever this is happening.

**Comment by Madouura** [comment]
>It looks like we do something special in stage-1 by manually splitting the devices to get them to work

AFAIK, this is not the case.
The only splitting done is to check each device in ``waitDevice`` separately, there is nothing altered beyond that one function.

**Comment by Slabity** [comment]
> AFAIK, this is not the case. The only splitting done is to check each device in `waitDevice` separately, there is nothing altered beyond that one function.

It does look like the root filesystem is mounted by the script instead of through systemd though, which could very well be why it works for people that use it for their root filesystem.

Unfortunately I have no way to test this myself other than trying to install a new copy of NixOS on it, but it's the only thing that I can see that's actually different from what everyone else here is doing. I'm going to try and see if there's a way to create a `systemd.mounts` entry that will search for each device individually instead of all at once. At least until systemd fixes the bug on their side.

**Comment by Slabity** [comment]
I made a PR that fixes the `mount.bcachefs.sh` script which should allow you to mount using the UUID instead of the devices themselves. I am able to now mount a non-root filesystem by setting `fsType = "bcachefs.sh"` like so:

```nix
"/data" = {
  device = "64adc9ee-d89d-4a2c-bb4d-6e22a7ab5219";
  fsType = "bcachefs.sh";
  options = [ "verbose" "nofail" "noatime" "x-systemd.device-timeout=10s" ];
};
```

**Comment by ziguana** [comment]
> Oddly I am actually still running into the original issue even on unstable. If I add the following:
> 
> ```nix
>     fileSystems."/data" = {
>       device = "/dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1";
>       fsType = "bcachefs";
>     };
> ```
> 
> Then systemd times out with `Timed out waiting for device /dev/nvme1n1p1:/dev/nvme2n1p1:/dev/md127p1` and then fails to boot.

same here. it sounds like we're waiting on https://github.com/systemd/systemd/issues/8234 ?

i see the `bcachefs.sh` workaround has been abandoned.

**Comment by Slabity** [comment]
> i see the `bcachefs.sh` workaround has been abandoned.

Sort of. I did not mean to close the PR altogether, but I forgot I had it still opened after 9 months when I deleted the repo.

The issue with that PR was that I couldn't guarantee it would work with it mounted as the root in stage-1 and nobody tested it since then to confirm what changes would be required to make it work. I can confirm the workaround does still work when mounting with systemd though if you want to add an override in your system to get it working.

I'd like to see if it's possible to get the mounting tool that's written in rust to work.
